helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java

```diff
@@ -0,0 +1,354 @@
+              inputMap.resourceFilter.addAll((List<String>) option.getValue());
+            } else if (optionKey.equals(OPTIONS_INSTANCE_FLT) && (option.getValue() instanceof List)) {
+              inputMap.instanceFilter.addAll((List<String>) option.getValue());
+            } else {
+              throw new InvalidParameterException(
+                  "Unsupported command or invalid format for Options : " + option);
+            }
+          }
+          break;
+        default:
+          throw new InvalidParameterException(
+              "Unsupported command for partitionAssignment : " + key);
+      }
+    }
+    return inputMap;
+  }
+
+  private ClusterState readClusterStateAndValidateInput(String clusterId, InputFields inputFields)
+      throws InvalidParameterException {
+
+    ClusterState clusterState = new ClusterState();
+    ConfigAccessor cfgAccessor = getConfigAccessor();
+    HelixDataAccessor dataAccessor = getDataAccssor(clusterId);
+    clusterState.resources = dataAccessor.getChildNames(dataAccessor.keyBuilder().idealStates());
+    // Add existing live instances and new instances from user input to instances list.
+    clusterState.instances = dataAccessor.getChildNames(dataAccessor.keyBuilder().liveInstances());
+    clusterState.instances.addAll(inputFields.newInstances);
+
+    // Check if to be removed instances and old instances in swap node exist in live instance.
+    if (!inputFields.nodeSwap.isEmpty() || !inputFields.instancesToRemove.isEmpty()) {
+      Set<String> liveInstanceSet = new HashSet<>(clusterState.instances);
+      for (Map.Entry<String, String> nodeSwapPair : inputFields.nodeSwap.entrySet()) {
+        if (!liveInstanceSet.contains(nodeSwapPair.getKey())) {
+          throw new InvalidParameterException("Invalid input: instance [" + nodeSwapPair.getKey()
+              + "] in SwapInstances does not exist in cluster.");
+        }
+      }
+      for (String instanceToRemove : inputFields.instancesToRemove) {
+        if (!liveInstanceSet.contains(instanceToRemove)) {
+          throw new InvalidParameterException("Invalid input: instance [" + instanceToRemove
+              + "] in RemoveInstances does not exist in cluster.");
+        }
+      }
+      if (!inputFields.instancesToRemove.isEmpty()) {
+        clusterState.instances.removeIf(inputFields.instancesToRemove::contains);
+      }
+    }
+
+    // Read instance and cluster config.
+    // It will throw exception if there is no instanceConfig for newly added instance.
+    for (String instance : clusterState.instances) {
+      InstanceConfig config = cfgAccessor.getInstanceConfig(clusterId, instance);
+      clusterState.instanceConfigs.add(config);
+    }
+    clusterState.clusterConfig = cfgAccessor.getClusterConfig(clusterId);
+    return clusterState;
+  }
+
+  private AssignmentResult computeOptimalAssignmentForResources(InputFields inputFields,
```

Comment:
Shall we move the logic into HelixUtil or RebalanceUtil? Because this could be something leveraged by Java API as well.

Reply:
HelixUtil already have multiple different APIs serving this feature. Also `InputFields` and `ClusterState` are private class and only have meaning in this context.

Reply:
@junkaixue This logic is needed only when we don't have the cluster data cache maintained for helix-rest. I believe adding cluster data cache in helix-rest is our next step. And after that, we can just run a rebalance pipeline instead of duplicating these logics here.
As I discussed with Xiaoyuan, the next step will take time. So we proposed to have this private method temporarily. But we should keep it private since it will be removed eventually.